juju/client/connection.py

```diff
@@ -22,49 +23,73 @@
-    'ModelGeneration': {'versions': [4]},
-    'ModelManager': {'versions': [9]},
-    'ModelUpgrader': {'versions': [1]},
-    'Payloads': {'versions': [1]},
-    'Pinger': {'versions': [1]},
-    'Resources': {'versions': [3]},
-    'SSHClient': {'versions': [4]},
-    'SecretBackends': {'versions': [1]},
-    'Secrets': {'versions': [1, 2]},
-    'Spaces': {'versions': [6]},
-    'Storage': {'versions': [6]},
-    'Subnets': {'versions': [5]},
-    'UserManager': {'versions': [3]},
+    'Action': [7],
+    'Admin': [3],
+    'AllModelWatcher': [4],
+    'AllWatcher': [3],
+    'Annotations': [2],
+    'Application': [17, 19],
+    'ApplicationOffers': [4],
+    'Backups': [3],
+    'Block': [2],
+    'Bundle': [6],
+    'Charms': [6],
+    'Client': [6, 7],
+    'Cloud': [7],
+    'Controller': [11],
+    'CredentialManager': [1],
+    'FirewallRules': [1],
+    'HighAvailability': [2],
+    'ImageMetadataManager': [1],
+    'KeyManager': [1],
+    'MachineManager': [10],
+    'MetricsDebug': [2],
+    'ModelConfig': [3],
+    'ModelGeneration': [4],
+    'ModelManager': [9],
+    'ModelUpgrader': [1],
+    'Payloads': [1],
+    'Pinger': [1],
+    'Resources': [3],
+    'SSHClient': [4],
+    'SecretBackends': [1],
+    'Secrets': [1, 2],
+    'Spaces': [6],
+    'Storage': [6],
+    'Subnets': [5],
+    'UserManager': [3],
 }
 
+# We don't generate code for these, as we can never use them.
+# The controller happily lists them though, don't warn about these.
+known_unsupported_facades = frozenset(("""
+ActionPruner Agent AgentLifeFlag AgentTools ApplicationScaler CAASAdmission CAASAgent
+CAASApplication CAASApplicationProvisioner CAASFirewaller CAASFirewallerSidecar
+CAASModelConfigManager CAASModelOperator CAASOperator CAASOperatorProvisioner
+CAASOperatorUpgrader CAASUnitProvisioner CharmDownloader CharmRevisionUpdater Cleaner
+CredentialValidator CrossModelRelations CrossModelSecrets Deployer DiskManager
+EntityWatcher EnvironUpgrader ExternalControllerUpdater FanConfigurer
+FilesystemAttachmentsWatcher Firewaller HostKeyReporter ImageMetadata InstanceMutater
+InstancePoller KeyUpdater LeadershipService LifeFlag LogForwarding Logger MachineActions
+MachineUndertaker Machiner MeterStatus MetricsAdder MetricsManager MigrationFlag
+MigrationMaster MigrationMinion MigrationStatusWatcher NotifyWatcher OfferStatusWatcher
+PayloadsHookContext Provisioner ProxyUpdater Reboot RelationStatusWatcher
+RelationUnitsWatcher RemoteRelationWatcher RemoteRelations ResourcesHookContext
+RetryStrategy SecretBackendsManager SecretBackendsRotateWatcher SecretsDrain
+SecretsManager SecretsRevisionWatcher SecretsTriggerWatcher Singular StatusHistory
+StorageProvisioner StringsWatcher Undertaker UnitAssigner Uniter UpgradeSeries
+UpgradeSteps Upgrader UserSecretsDrain UserSecretsManager VolumeAttachmentPlansWatcher
+VolumeAttachmentsWatcher
+""" + """
+CrossController MigrationTarget ModelSummaryWatcher NotifyWatcher
+""").strip().split())
```

Comment:
```suggestion
known_unsupported_facades = frozenset((
    'ActionPruner',
    'Agent',
    'AgentLifeFlag',
    'AgentTools',
    'ApplicationScaler',
    'CAASAdmission',
    'CAASAgent',
    'CAASApplication',
    'CAASApplicationProvisioner',
    'CAASFirewaller',
    'CAASFirewallerSidecar',
    'CAASModelConfigManager',
    'CAASModelOperator',
    'CAASOperator',
    'CAASOperatorProvisioner',
    'CAASOperatorUpgrader',
    'CAASUnitProvisioner',
    'CharmDownloader',
    'CharmRevisionUpdater',
    'Cleaner',
    'CredentialValidator',
    'CrossController',
    'CrossModelRelations',
    'CrossModelSecrets',
    'Deployer',
    'DiskManager',
    'EntityWatcher',
    'EnvironUpgrader',
    'ExternalControllerUpdater',
    'FanConfigurer',
    'FilesystemAttachmentsWatcher',
    'Firewaller',
    'HostKeyReporter',
    'ImageMetadata',
    'InstanceMutater',
    'InstancePoller',
    'KeyUpdater',
    'LeadershipService',
    'LifeFlag',
    'LogForwarding',
    'Logger',
    'MachineActions',
    'MachineUndertaker',
    'Machiner',
    'MeterStatus',
    'MetricsAdder',
    'MetricsManager',
    'MigrationFlag',
    'MigrationMaster',
    'MigrationMinion',
    'MigrationStatusWatcher',
    'MigrationTarget',
    'ModelSummaryWatcher',
    'NotifyWatcher',
    'NotifyWatcher',
    'OfferStatusWatcher',
    'PayloadsHookContext',
    'Provisioner',
    'ProxyUpdater',
    'Reboot',
    'RelationStatusWatcher',
    'RelationUnitsWatcher',
    'RemoteRelationWatcher',
    'RemoteRelations',
    'ResourcesHookContext',
    'RetryStrategy',
    'SecretBackendsManager',
    'SecretBackendsRotateWatcher',
    'SecretsDrain',
    'SecretsManager',
    'SecretsRevisionWatcher',
    'SecretsTriggerWatcher',
    'Singular',
    'StatusHistory',
    'StorageProvisioner',
    'StringsWatcher',
    'Undertaker',
    'UnitAssigner',
    'Uniter',
    'UpgradeSeries',
    'UpgradeSteps',
    'Upgrader',
    'UserSecretsDrain',
    'UserSecretsManager',
    'VolumeAttachmentPlansWatcher',
    'VolumeAttachmentsWatcher',
))
```

Reply:
I was trying to avoid that... maybe I can just stash this long list into another file?

Reply:
Could put `excluded_facade_versions`, `client_facades`, and `known_unsupported_facades` in their own module. Gives us a place to put a nice docstring explaining what their deal is too.

Reply:
moved to a new file